### PR TITLE
Enable abl mesh/grid auto restore

### PR DESCRIPTION
**Pull request by Bob-the-Kuhn**
I didn’t realize all the activity in this area by the heavyweights until I was 99% done.
Here’s my ideas on how to make saving & restoring bed levelling settings bullet proof.  Use them if they fit in.

==================================================

This is all independent of Mesh Leveling.  None of the Mesh leveling code was touched.

I’ve created a flag to signify if there is a valid ABL bed correction matrix/grid.  
- G29 sets it to true.
- G28 uses it to determine if ABL leveling should be turned on.  G28 no longer automatically exits with ABL disabled.
- M420 uses it when S1 and V parameters are specified.

Configuration_store uses the flag along with the leveling_is_on variable to indicate if a saved matrix/grid is valid or not.
- leveling_is_on has been changed from a Boolean flag for Mesh Leveling to a variable that indicates if a valid mesh/matrix/grid has been stored.
- The restore process looks at leveling_is_on  and, if it matches the currently active leveling system, sets the flag to valid.  G28 will then set leveling active.

I’ve modified M420 so that ABL is enabled only if G28 has been run and if the flag says the grid/mesh is valid.  The display of the grid/mesh is disabled if the flag is set to invalid.

### CODE REVIEW

**Comment by Bob-the-Kuhn** [comment]
Strange - I wasn't able to fully squash this branch.  Please ignore the first commit **all code changes are in Marlin_main.cpp** .   There are two files that changed.

**Comment by Roxy-3D** [comment]
I've been ignoring the discussion on saving Auto Bed Leveling state because it is fully solved in the UBL branch.   The printer resets...  and it knows how to correct itself.     The user can optionally have the bed leveling turned on or off after reset.   (My printers turn on and the bed leveling is fully active.)

**Comment by Bob-the-Kuhn** [comment]
Don't you need/want to go through a homing cycle before turning on bed leveling?

**Comment by Roxy-3D** [comment]
> Don't you need/want to go through a homing cycle before turning on bed leveling?

The X and Y are homed independently.   And during homing, you have the limit switches so you can set position with confidence.   The Z is a little different story for several reasons.  Depending upon where you choose to have the nozzle when you home Z, you can get slightly different results.   First because the bed can be tilted.   But also, imperfections in the flatness of the bed can enter in.

To answer your question...   Its actually the opposite.   In the UBL code, Bed Level Corrections are turned off during homing.   And then the state you entered G28 is restored at the end.

**Comment by Bob-the-Kuhn** [comment]
Yes, G28 turns off leveling, does its thing & then enables leveling if appropriate.

I just assumed that we didn't want leveling enabled before G28.

I don't see any danger in turning on leveling before doing a G28.  Before G28 you don't know where you're at.  Enabling leveling before G28 just means you still don't know where you're at.

Does UBL have a means of guarding against the following init scenarios?
- switching from one leveling system to another would/could result in an invalid mesh/grid/matrix being used
- the user never saved a mesh/grid/matrix to the EEPROM so garbage would be restored at init

I'll wander over to UBL & see what's been done.  It's a good idea to only have one system.

**Comment by Bob-the-Kuhn** [comment]
I just downloaded DEV-UBL and looked at configuration_store.

It is only concerned with UBL (understandable) and it saves the valid/invalid state of the mesh.

I don't see that the UBL system impacts what I'm proposing (other than the enable before/after G28 assumption).  When UBL & RCBugFix is merged there'll have to be a change to the EEPROM system no matter what.  In the unified EEPROM system I'm hoping that each bed leveling system will have its own valid/invalid flag.

I'm OK with enabling leveling before G28.

I'll wait a couple of days to see if anyone else weighs in on the subject before modifying the code.

File: Marlin/configuration_store.cpp [hunk 1]
```diff
@@ -511,7 +535,7 @@ void Config_Postprocess() {
       // Mesh (Manual) Bed Leveling
       //
 
-      bool leveling_is_on;
+      uint8_t leveling_is_on;
```
Comment (thinkyhead) [inline Marlin/configuration_store.cpp:538]: This instance will conceal the previous instance on line 475.

File: Marlin/configuration_store.cpp [hunk 2]
```diff
@@ -260,14 +261,36 @@ void Config_Postprocess() {
...
+    #endif
+    
     //
     // Mesh Bed Leveling
     //
 
     #if ENABLED(MESH_BED_LEVELING)
       // Compile time test that sizeof(mbl.z_values) is as expected
       typedef char c_assert[(sizeof(mbl.z_values) == (MESH_NUM_X_POINTS) * (MESH_NUM_Y_POINTS) * sizeof(dummy)) ? 1 : -1];
-      const bool leveling_is_on = TEST(mbl.status, MBL_STATUS_HAS_MESH_BIT);
```
Comment (thinkyhead) [inline Marlin/configuration_store.cpp:270]: For MBL I believe we do want to save whether or not it has a valid mesh.

File: Marlin/configuration_store.cpp [hunk 3]
```diff
@@ -260,14 +261,36 @@ void Config_Postprocess() {
...
+ *
+ *  leveling_is_on
+ *    0 - leveling is off or the current ABL grid/matrix is invalid
+ *    1 - valid Mesh Leveling 
+ *    2 - valid AUTO_BED_LEVELING_3POINT correction grid
+ *    3 - valid AUTO_BED_LEVELING_LINEAR correction grid 
+ *    4 - valid AUTO_BED_LEVELING_BILINEAR matrix
+ */
+    
+
```
Comment (thinkyhead) [inline Marlin/configuration_store.cpp:275]: Some extra spaces and newlines fell into the code here.

File: Marlin/Marlin_main.cpp [hunk 4]
```diff
@@ -3419,7 +3419,17 @@ inline void gcode_G4() {
...
+ * Bed leveling
+ *  If there is a valid bed leveling grid then bed leveling is
+ *  turned on after homing is completed - even if it wasn't on
+ *  before G28 was executed.  A valid grid is created either by
+ *  G29 or by M501.  Bed leveling using M501 grids is only
+ *  enabled by the G28 command. Bed leveling using G29 grids is
+ *  enabled within the G29 command.
  */
+ 
+extern bool ABL_grid_valid;
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:3431]: "`abl_grid_valid`" (all lowercase) is preferred for variable names.

**Comment by thinkyhead** [comment]
I think the logic should be that if there is a valid mesh/grid in the EEPROM then —yes— it should be loaded with `M500` and the "has valid mesh" state should be noted. However, I disagree with re-enabling leveling automatically after `G28` as a default behavior. This extends `G28` beyond its single purpose, which is to establish a known position (_disabling_ leveling in the process). We have `M420` to enable bed leveling, so that should be inserted after `G28` if users want to do that.

**Comment by Bob-the-Kuhn** [comment]
When to enable the mesh/grid that is read from EEPROM ... definitely needs some discussion.

@Roxy-3D thinks it should be like the UBL system
- enable auto bed leveling if a valid mesh/grid is read from the EEPROM
- G28 restores auto bed leveling to what ever state it was in before G28 was invoked.  That's what RCBugFix does with MESH leveling right now.

I can easily live with enabling the MESH/GRID restored from EEPROM immediately.  My preference is to wait until G28 has been run but that's not how UBL works and that's not how MESH bed leveling.

What I would oppose is G28 turning off ABL but **NOT** restoring it at the end.  I see no value add to making the user have to enable ABL every time G28 is run.

@thinkyhead  - is the UBL system acceptable to you?  If yes then I'll make those changes

**Comment by Roxy-3D** [comment]
> Does UBL have a means of guarding against the following init scenarios?
> switching from one leveling system to another would/could result in an invalid mesh/grid/matrix being used the user never saved a mesh/grid/matrix to the EEPROM so garbage would be restored at init
I'll wander over to UBL & see what's been done. It's a good idea to only have one system.

Here is a very simple case worthy of a  quick 'Thought Experiment'.  Suppose the user has the machine homed and doing corrections in UBL.    
- The user successfully performs a print.   
- The user takes the glass off the bed to put into the freezer to release the print.
- The user takes another already calibrated piece of glass and puts it on the printer.
- The user loads the correction matrix for that piece of glass via a G29 L 2   (assuming that Mesh is in slot 2)
- The user does a G28 to re-home the machine for the new piece of glass.
- The machine is back to correcting for the piece of glass that is on the machine.
- The user is ready to do another print.

This sequence will make sens for a lot of users.   And in fact...  It is what I'm doing.   I want the machine to always be correcting for the piece of glass clipped on the bed.

I don't have any G29 stuff in my Startup GCode produced by the Slicer.  I don't need it.  The printer wakes up loads Mesh number 1.    If I end up using a different piece of glass, I manually tell the printer to do a G29 L2 to load the mesh for my 2nd piece of glass.   That works well for me.

>  However, I disagree with re-enabling leveling automatically after G28 as a default behavior. This extends G28 beyond its single purpose, which is to establish a known position (disabling leveling in the process). 

In the UBL code base, the default is not to automatically activate the bed leveling corrections.  The user has to take explicit action to make that the default case.   But with that said, doing a G28 to re-establish where the top surface of the glass is seems independent of what corrections are needed to follow the surface of the glass.   It seems to me leaving the bed leveling system in the same state (active or not active) across the G28 is a nice convenience for the user.

**Comment by Bob-the-Kuhn** [comment]
I hadn't considered having multiple valid mesh/grids in EEPROM for a given leveling system.

Definitely not supported by the current RCBugFix EEPROM layout.

I'd prefer that this discussion be put off until UBL is being merged into the main branch.

**Comment by Bob-the-Kuhn** [comment]
As long as we're messing with the EEPROM - how about adding PR #5762 into this PR?  

Makes sense to do the EEPROM changes all at the same time. 

On the other hand it would result in the merged PR having multiple purposes/functions.  It's nice to keep a PR as focused as possible.

**Comment by Roxy-3D** [comment]
> I hadn't considered having multiple valid mesh/grids in EEPROM for a given leveling system.
> Definitely not supported by the current RCBugFix EEPROM layout.

Its already fully supported in the UBL code base.   And it is worth while looking at how UBL does a full CRC to guarantee the integrity of the EEPROM data.   Doing a checksum of the data is a step in the right direction.  But it won't protect against multi-bit problems.   For that, you need to do a CRC.   That code can be lifted directly out of UBL and applied.   The checksum routines would just get replaced by the CRC routines.  (And they are very small routines!)

**Comment by Bob-the-Kuhn** [comment]
@Roxy-3D & @thinkyhead 

How about the following?
1. If the UBL/RCBugFix merger is less than 3 months away we just close this PR.
2. If longer then I make the following changes to be consistent with the UBL methods:
- only enable grid/correction matrix restored from EEPROM with M420 command
- Make G28 so it restores the bed leveling state to what it was when G28 was kicked off
3. I roll the PR #5762 changes into this PR (make all the EEPROM changes at the same time)
4. Mesh bed leveling stays as is in the latest RCBugFix

**Comment by Bob-the-Kuhn** [comment]
This PR is obsolete.  The UBL system has a slightly different philosophy.

**Pull request closed without merging**